### ISSUE TITLE
Thanks and ASIO question

### ISSUE THREAD

**Issue by romange** [boostorg/beast#1391]
Hi Vinnie,

I just wanted to tell thanks for your library. I've integrated it into a framework which I've built over ASIO with Fibers to provide HTTP (beast) and RPC backend with IOContext-per-thread threading model. The code is asynchronous thanks to Boost.Fibers scheduler which allows me to write asynchronous code while preserving synchronous code style (no callbacks, no futures). The framework is very efficient and allows me to reach 300,000 QPS per core (1M qps on my laptop with 4 cores, for example). Anyway, if you are curious it's [here](https://github.com/romange/gaia). 

I have a question about ASIO sockets, maybe you know the answer. It seems that the only way to learn when socket is in disconnected/error state is to call some action on it and get a result. But suppose I have a client socket that got disconnected 5 mins before. For the last 5mins my process did not need to do anything on this socket so it does not realize that it's disconnected. Only when it writes something it gets an error_code and can start reconnecting. Is there a way to get a signal from Asio socket earlier? If not, do you know how to workaround this issue and be always up-to-date regarding the state of the socket?

**Comment by djarek** [comment]
You can perform `socket.async_wait(boost::asio::ip::tcp::socket::wait_read, wait_handler);`, which will signal readiness if the socket becomes "readable" (either the socket has pending data or it has been closed).

**Comment by romange** [comment]
Interesting! Could it be that wait_error is the better option for this scenario? In any case, you gave me the direction and I will check it out!

**Comment by vinniefalco** [comment]
> For the last 5mins my process did not need to do anything on this socket so it does not realize that it's disconnected.

I think you should always have a read pending, so that you can get timely notification of `error::eof` (that's TCP/IP shutdown's FIN packet). Or if you have an SSL stream then you can detect the protocol-level close_notify (also delivered via `error::eof`).

**Comment by romange** [comment]
Thanks @vinniefalco  and @djarek  I rewrote my code to always have an outstanding pending read.
In order to differentiate between EOF or valid read signal I have to follow up with `receive()`. 
I had 2 options:
1. To use MESSAGE_PEEK for 1 byte just for sake of checking for EOF. Advantage: the control over socket reads stays in the upper layers outside of my socket wrapper but it always knows the correct state of the socket. The disadvantage: could be twice more system calls per read. In the extreme case, for each notification from async_wait, we issue MESSAGE_PEEK and then the real call from upper level calls the `read()` as well.    
2. Use the "real" `read()` in the wrapper to fetch the data optimistically into intermediate buffer even though noone from upper layer is interested yet in this data. Disadvantage: more complicated state machine in my code + more memory copies.
I chose option (2) assuming that system calls are much more expensive that copying small blocks of memory.
